### DISCUSSION TITLE
Does vitest have the same problem as jest in this regard?

### DISCUSSION THREAD

**Discussion by spicylemonhaha** [vitest-dev/vitest#1665] [Q&A]
When we analysed our customer’s tests, we soon noticed a recurring problem: that the most significant impact on performance was the number of modules which needed to be imported by the tests in question. Jest creates a new [module registry](https://jestjs.io/docs/configuration#resetmodules-boolean) for each test file and has to crawl through the whole dependency chain of modules which are imported by the test suite, even if they are completely unused. It is easy for developers to overlook this fact given that build tools such as webpack can eliminate dead code via [tree-shaking](https://webpack.js.org/guides/tree-shaking/), but this is only the case for production builds and does not help with test execution or development servers.

https://blog.codecentric.de/en/2021/11/javascript-test-performance-getting-the-best-out-of-jest/

**Comment by sheremet-va** [comment]
Yes, Vitest loads every imported module. It won't load a module, if imported variable is not used tho, which is a rare case and usually handled by linters.

I don't think tree-shaking is possible with Node.js context tho, we would need to bundle code for each test file.

**Reply by spicylemonhaha** [reply]
That's right, I'm doing a test framework research for our department. After my test, I found that vitest is faster than jest, but I want to know why, this will help me introduce this framework to the group. I originally thought vitest didn't have this problem, so it would be faster than jest in the first test of the entire code. But that doesn't seem to be the reason now. So begging you to tell me why vitest is almost twice as fast as jest when testing some of the same use cases. If possible, could you please explain or show the relevant source code concisely.

**Reply by Demivan** [reply]
What do you use now to transform your code with Jest? If you are using Babel that is your answer. Babel is slow.

**Reply by spicylemonhaha** [reply]
> What do you use now to transform your code with Jest? If you are using Babel that is your answer. Babel is slow.

But both my jest and vitest are tested in the same environment and both have babel

**Reply by Demivan** [reply]
You used babel plugin for Vite? Because Vite uses esbuild by default.

**Reply by spicylemonhaha** [reply]
https://github.com/spicylemonhaha/test_example  In this example, you will find that vitest performs better than jest (my environment is mac m1 pro)

**Reply by sheremet-va** [reply]
> https://github.com/spicylemonhaha/test_example In this example, you will find that vitest performs better than jest (my environment is mac m1 pro)

According to your example, you don't use babel in Vitest. We use Vite to transform files, and it doesn't use babel, if not explicitly specified with vite-plugin-babel.

**Reply by spicylemonhaha** [reply]
I have now removed babel from the demo and changed the module syntax to commonjs (the reason why I use babel in jest is to use ESM), but vitest is still much faster than jest after testing, so what is the possible reason? , as far as I know jest is also multithreaded
